### PR TITLE
Board: Add MoveCard action

### LINKED ISSUE #1297: Card: Move
Should be triggered from actions menu + from State name on the EditCard component

### PR DESCRIPTION
**Pull request by Dvinyanin**
Signed-off-by: Dvinyanin Alexandr [dvinyanin.alexandr@gmail.com](mailto:dvinyanin.alexandr@gmail.com)

### CODE REVIEW

File: plugins/board-assets/lang/ru.json [hunk 1]
```diff
@@ -53,6 +53,12 @@
     "NewListPlaceholder": "Введите заголовок колонки...",
     "AddACard": "Добавить карточку",
     "AddCard": "Добавить карточку",
-    "CardTitlePlaceholder": "Введите заголовок для этой карточки..."
+    "CardTitlePlaceholder": "Введите заголовок для этой карточки...",
+    "MoveCard": "Переместить карточку",
+    "SelectDestination": "Выберите цель",
```
Comment (annano) [inline plugins/board-assets/lang/ru.json]: It is better to translate like `Выберите список` and in general probably we should have an agreed translation for `list` and use it everywhere.

File: plugins/board-resources/src/components/MoveCard.svelte [hunk 2]
```diff
@@ -0,0 +1,124 @@
+<script lang="ts">
```
Comment (annano) [inline plugins/board-resources/src/components/MoveCard.svelte]: If it is a popup, should add a `popups` folder and move it there

File: plugins/board-resources/src/components/MoveCard.svelte [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+      on:click={() => {
+        dispatch('close')
+      }}
+    />
+  </div>
+</div>
+
+<style lang="scss">
+  .container {
+    display: flex;
```
Comment (annano) [inline plugins/board-resources/src/components/MoveCard.svelte]: Can we reuse global styles from `themes/packages/styles/_layouts.scss` as much as possible here?

File: plugins/board-resources/src/components/move-selectors/SpaceSelect.svelte [hunk 4]
```diff
@@ -0,0 +1,26 @@
+<script lang="ts">
+    import { IntlString, translate } from '@anticrm/platform'
```
Comment (annano) [inline plugins/board-resources/src/components/move-selectors/SpaceSelect.svelte]: I am not sure, but maybe it is better to create a folder `selectors` and move it there?

File: plugins/board-resources/src/index.ts [hunk 5]
```diff
@@ -22,8 +24,13 @@ import EditCard from './components/EditCard.svelte'
 import KanbanCard from './components/KanbanCard.svelte'
 import TemplatesIcon from './components/TemplatesIcon.svelte'
 import KanbanView from './components/KanbanView.svelte'
+import MoveView from './components//MoveCard.svelte'
 import { addCurrentUser, canAddCurrentUser, isArchived, isUnarchived } from './utils/CardUtils'
 
+async function moveCard (object: Card): Promise<void> {
```
Comment (annano) [inline plugins/board-resources/src/index.ts]: `showMoveCardPopup` ?

**Review by haiodo — CHANGES_REQUESTED** [review]
Collections on move

File: plugins/board-assets/lang/ru.json (hunk 1 above)
Comment (Dvinyanin) [reply plugins/board-assets/lang/ru.json]: Done

File: plugins/board-resources/src/components/MoveCard.svelte (hunk 3 above)
Comment (Dvinyanin) [reply plugins/board-resources/src/components/MoveCard.svelte]: Done

File: plugins/board-resources/src/components/move-selectors/SpaceSelect.svelte (hunk 4 above)
Comment (Dvinyanin) [reply plugins/board-resources/src/components/move-selectors/SpaceSelect.svelte]: Done

File: plugins/board-resources/src/index.ts (hunk 5 above)
Comment (Dvinyanin) [reply plugins/board-resources/src/index.ts]: Renamed

File: plugins/attachment-resources/src/utils.ts [hunk 6]
```diff
@@ -84,3 +86,19 @@ export function getType (type: string): 'image' | 'video' | 'audio' | 'pdf' | 'o
 
   return 'other'
 }
+
+export async function moveAttachments (doc: Doc, space: Ref<Space>): Promise<void> {
+  const client = getClient()
+  const hierarchy = client.getHierarchy()
+  const attributes = hierarchy.getAllAttributes(doc._class)
+  for (const [name, attribute] of attributes) {
```
Comment (haiodo) [inline plugins/attachment-resources/src/utils.ts]: AttachedDoc is not an attachment* plugin concept, you should move this logic into package/core

File: plugins/attachment-resources/src/utils.ts [hunk 7]
```diff
@@ -84,3 +86,19 @@ export function getType (type: string): 'image' | 'video' | 'audio' | 'pdf' | 'o
...
+    if (hierarchy.isDerived(attribute.type._class, core.class.Collection)) {
+      const collection = attribute.type as Collection<AttachedDoc>
+      const allAttached = await client.findAll(collection.of, { attachedTo: doc._id })
+      for (const attached of allAttached) {
+        moveAttachments(attached, space).catch((err) => console.log('failed to move', name, err))
```
Comment (haiodo) [inline plugins/attachment-resources/src/utils.ts]: It is bad way to mix async and promise concepts, please use async approach during one function code

Comment (Dvinyanin) [reply plugins/attachment-resources/src/utils.ts]: Fixed

File: plugins/attachment-resources/src/utils.ts (hunk 6 above)
Comment (Dvinyanin) [reply plugins/attachment-resources/src/utils.ts]: Not sure but moved to `TxOperations`
Should we update only `space` for AttachedDoc?

File: packages/core/src/operations.ts [hunk 8]
```diff
@@ -183,4 +183,21 @@ export class TxOperations implements Omit<Client, 'notify'> {
     }
     return this.removeDoc(doc._class, doc.space, doc._id)
   }
+
+  async updateWithAttachments<T extends Doc>(doc: T, update: DocumentUpdate<T>, retrieve?: boolean): Promise<void> {
```
Comment (haiodo) [inline packages/core/src/operations.ts]: Please move to server/storage/ as trigger for TxUpdateDoc, similar to processRemove.

Comment (Dvinyanin) [reply packages/core/src/operations.ts]: Done

File: server/core/src/storage.ts [hunk 9]
```diff
@@ -333,6 +333,25 @@ class TServerStorage implements ServerStorage {
...
+    const rtx = actualTx as TxUpdateDoc<Doc>
+    if (rtx.operations.space === undefined || rtx.operations.space === rtx.objectSpace) return []
+    const result: Tx[] = []
+    const factory = new TxFactory(core.account.System)
+    for (const [, attribute] of this.hierarchy.getAllAttributes(rtx.objectClass)) {
+      if (!this.hierarchy.isDerived(attribute.type._class, core.class.Collection)) continue
+      const collection = attribute.type as Collection<AttachedDoc>
+      const allAttached = await this.findAll(ctx, collection.of, { attachedTo: rtx.objectId, space: rtx.objectSpace })
+      const allTx = allAttached.map(({ _class, space, _id }) =>
+        factory.createTxUpdateDoc(_class, space, _id, { space: rtx.operations.space }, rtx.retrieve)
```
Comment (haiodo) [inline server/core/src/storage.ts]: I think we do not need to pass retrieve for attached here.

Comment (haiodo) [inline server/core/src/storage.ts]: You need to ignore attached with space not same as previous doc value.

**Review by haiodo — CHANGES_REQUESTED** [review]

File: server/core/src/storage.ts (hunk 9 above)
Comment (Dvinyanin) [reply server/core/src/storage.ts]: Space is specified in `findAll`, can it return attached with different space?

Comment (Dvinyanin) [reply server/core/src/storage.ts]: Removed

File: plugins/board-resources/src/components/popups/MoveCard.svelte [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+        dispatch('close')
+      }}
+    />
+    <Button
+      label={board.string.Move}
+      size={'small'}
+      disabled={status !== OK || object.state === selected.state && object.rank === selected.rank}
+      kind={'primary'}
+      on:click={() => {
+        client.update(object, selected)
```
Comment (annano) [inline plugins/board-resources/src/components/popups/MoveCard.svelte]: Lets update only properties which have changed

Comment (Dvinyanin) [reply plugins/board-resources/src/components/popups/MoveCard.svelte]: Done

File: server/core/src/storage.ts [hunk 11]
```diff
@@ -337,6 +337,25 @@ class TServerStorage implements ServerStorage {
...
+    const actualTx = this.extractTx(tx)
+    if (!this.hierarchy.isDerived(actualTx._class, core.class.TxUpdateDoc)) return []
+    const rtx = actualTx as TxUpdateDoc<Doc>
+    if (rtx.operations.space === undefined || rtx.operations.space === rtx.objectSpace) return []
+    const result: Tx[] = []
+    const factory = new TxFactory(core.account.System)
+    for (const [, attribute] of this.hierarchy.getAllAttributes(rtx.objectClass)) {
+      if (!this.hierarchy.isDerived(attribute.type._class, core.class.Collection)) continue
+      const collection = attribute.type as Collection<AttachedDoc>
+      const allAttached = await this.findAll(ctx, collection.of, { attachedTo: rtx.objectId, space: rtx.objectSpace })
```
Comment (haiodo) [inline server/core/src/storage.ts:350]: We need to call refactor buildRemovedDoc to allow find previous space value for document, and find only attached documents with it's space value.

Comment (haiodo) [reply server/core/src/storage.ts:350]: Discussed, resolved

**Review by haiodo — APPROVED** [review]

**Pull request merged**